chia/wallet/wallet_protocol.py

```diff
@@ -0,0 +1,63 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, List, Optional, Set
+
+from blspy import G1Element
+from typing_extensions import Protocol
+
+from chia.server.ws_connection import WSChiaConnection
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.ints import uint8, uint32, uint64, uint128
+from chia.wallet.wallet_coin_record import WalletCoinRecord
+
+if TYPE_CHECKING:
+    from chia.wallet.wallet_state_manager import WalletStateManager
+
+
+class WalletProtocol(Protocol):
+
+    # TODO: it seems like this should return WalletType instead
+    @classmethod
+    def type(cls) -> uint8:
+        ...
+
+    def id(self) -> uint32:
+        ...
+
+    async def coin_added(self, coin: Coin, height: uint32, peer: WSChiaConnection) -> None:
+        ...
+
+    async def get_new_puzzlehash(self) -> bytes32:
+        ...
+
+    async def select_coins(
+        self,
+        amount: uint64,
+        exclude: Optional[List[Coin]] = None,
+        min_coin_amount: Optional[uint64] = None,
+        max_coin_amount: Optional[uint64] = None,
+    ) -> Set[Coin]:
+        ...
+
+    async def get_confirmed_balance(self, record_list: Optional[Set[WalletCoinRecord]] = None) -> uint64:
+        ...
+
+    async def get_unconfirmed_balance(self, unspent_records: Optional[Set[WalletCoinRecord]] = None) -> uint128:
+        ...
+
+    async def get_spendable_balance(self, unspent_records: Optional[Set[WalletCoinRecord]] = None) -> uint128:
+        ...
+
+    async def get_pending_change_balance(self) -> uint64:
+        ...
+
+    async def get_max_send_amount(self, unspent_records: Optional[Set[WalletCoinRecord]] = None) -> uint128:
+        ...
+
+    def puzzle_hash_for_pk(self, pubkey: G1Element) -> bytes32:
+        ...
+
+    # pylint says:
+    # E0601: Using variable 'WalletStateManager' before assignment (used-before-assignment)
```

Comment:
I think explaining why this is ok might be more useful?

```suggestion
    # WalletStateManager is only imported for type hinting thus leaving pylint unable to process this
```

Reply:
that's closer to an explanation. But the error message from pylint is still nonsensical. There is no "use" nor "assignment".

Reply:
or is pylint referring to the *type* `WalletStateManager` as a "variable"? and the definition of that type as "assignment"?

Reply:
Yep, you're using `WalletStateManager` before pylint knows what it is since it is only conditionally imported due to `if typing.TYPE_CHECKING:`.  In some sense it is useful to think of Python as not having variables.  You have names that reference objects and that's about it.  As distinct from, say, C variables where an int variable is a block of memory holding the int.  Functions are objects.  As are classes, modules, methods, stack frames, etc.